content/posts/vault-migration/index.md

```diff
@@ -0,0 +1,558 @@
+  ```
+
+### 3. Block access to the old Vault instance
+
+Before creating the snapshot, it’s important to prevent any writes to the old
+Vault instance. This step ensures no data is changed after the snapshot and
+avoids issues with lease revocation.
+
+- Backup and delete the ingress for the old Vault instance
+
+  ```bash
+  kubectl get ing vault -o yaml > "${BACKUPS_DIR}/${OLD_VAULT_INGRESS_FILENAME}"
+
+  kubectl delete ing vault --wait=true
+  ```
+
+### 4. Create snapshot of the old Vault instance
+
+We now take a snapshot of the Vault data using the Raft backend's built-in
+snapshot mechanism. After the snapshot is saved and downloaded, we immediately
+disable the old Vault to avoid unintended behavior.
+
+- Log in with the migration token
+
+  ```bash
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    vault login "${OLD_VAULT_ADMIN_TOKEN}"
+  ```
+
+- Create the snapshot in the leader pod
+
+  ```bash
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    vault operator raft snapshot save "${snapshot_filepath}"
+  ```
+
+- Dowload the snapshot
+
+  ```bash
+  kubectl cp "${OLD_VAULT_NAME}-${leader_index}":"${snapshot_filepath}" \
+    "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}"
+
+  kubectl exec -it "${OLD_VAULT_NAME}-${leader_index}" -- \
+    rm "${snapshot_temp_filepath}"
+  ```
+
+- Disable the old Vault instance by changing the command of the Vault
+  StatefulSet to `sleep`
+
+  ```bash
+  kubectl get statefulset vault -o yaml > \
+  "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}"
+  cp "${BACKUPS_DIR}/${OLD_VAULT_STATEFULSET_FILENAME}" .
+
+  ed "${OLD_VAULT_STATEFULSET_FILENAME}" <<EOF
+  /args:/
+  +1,/-config=/c
+          - sleep 999999
+  .
+  w
+  q
+  EOF
+
+  kubectl apply -f "${OLD_VAULT_STATEFULSET_FILENAME}"
+  ```
+
+- Restart all pods to pick up the patched configuration
+  ```bash
+  kubectl delete pod --selector=app.kubernetes.io/name=vault
+  ```
+
+### 5. Restore the snapshot in the new Vault instance
+
+With the snapshot from the old instance in hand, we restore it to the new Vault
+instance. This effectively brings the new Vault cluster to the same data state
+as the old one.
+
+- Upload the snapshot the leader pod
+
+  ```bash
+  kubectl cp "${BACKUPS_DIR}/${OLD_VAULT_SNAPSHOT_FILENAME}" \
+    "vault-${leader_index}":"${snapshot_tmp}"
+  ```
+
+- Log in with the migration token
+
+  ```bash
+  kubectl exec -it "vault-${leader_index}" -- \
+    vault login "${NEW_VAULT_MIGRATION_TOKEN}"
+  ```
+
+- Run the restore command
+  ```bash
+  kubectl exec -it "vault-${leader_index}" -- \
+    sh -c "vault operator raft snapshot restore -force \"${snapshot_tmp}\""
+  ```
+
+### 6. Configure the new Vault cluster to use the old unseal key
+
+To allow the restored Vault data to function properly, we need to configure the
+new instance to use the same unseal key as the old Vault. This step modifies the
+config map and restarts the pods to apply the change.
```

Comment:
```suggestion
The restored state must be unsealed with the unseal key stored in the KeyVault associated with the _old_ Vault instance. We now need to unseal the state but store a new unseal key in the KeyVault associated with the _new_ Vault instance. This step modifies the config map by replacing the coordinates to the _new_ instance unseal key with those of the _old_ instance unseal key and restarts the pods to apply the change. After the restart, the pods should unseal the state automatically since they have access to the original unseal key.
```